File: 1888-minimum-number-of-flips-to-make-the-binary-string-alternating/1888-minimum-number-of-flips-to-make-the-binary-string-alternating.go
func minFlips(s string) int {
    n := len(s)
    
    s2 := s + s

    o := make([]byte, 2*n) // alternating string, start with 1
    z := make([]byte, 2*n) // alternating string, start with 0
    for i := 0; i < 2*n; i++ {
        if i % 2 == 0 {
            o[i] = '1'
            z[i] = '0'
        } else {
            z[i] = '1'
            o[i] = '0'
        }
    }

    ans := math.MaxInt32
    on, zn := 0, 0
    left, right := 0, 0
    for right < 2 * n {
        if s2[right] != o[right] {
            on++
        }
        if s2[right] != z[right] {
            zn++
        }

        if right - left + 1 > n {
            if s2[left] != o[left] {
                on--
            }
            if s2[left] != z[left] {
                zn--
            }
            left++

            ans = min(ans, min(on, zn))
        }

        right++
    }

    return ans
}

// sliding window, string
// time: O(n)
// space: O(n)

File: 1888-minimum-number-of-flips-to-make-the-binary-string-alternating/1888-minimum-number-of-flips-to-make-the-binary-string-alternating.py
class Solution:
    def minFlips(self, s: str) -> int:
        n = len(s)

        s2 = s + s

        o = '10' * n # alternating string start with 1 size 2 * n
        z = '01' * n # alternating string start with 0 size 2 * n

        ans = 10**10
        on, zn = 0, 0
        left, right = 0, 0
        while right < n * 2:
            if s2[right] != o[right]:
                on += 1
            if s2[right] != z[right]:
                zn += 1

            if right - left + 1 > n:
                if s2[left] != o[left]:
                    on -= 1
                if s2[left] != z[left]:
                    zn -= 1
                left += 1

                ans = min(ans, min(on, zn))
            
            right += 1

        return ans

# sliding window, string
# time: O(n)
# space: O(n)